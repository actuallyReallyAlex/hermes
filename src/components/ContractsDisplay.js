import React from 'react'
import PropTypes from 'prop-types'
import { Box, Heading, Text, Button } from 'grommet'
import { Checkmark, Close } from 'grommet-icons'
import { connect } from 'react-redux'
import { resetContract, setContract } from '../redux/actions/user'

/**
 * Displays a set of item contracts.
 */
const ContractsDisplay = ({
  contracts,
  currentContract,
  handleResetContract,
  handleSetContract
}) => {
  return (
    <Box gap="small" margin={{ vertical: 'medium' }}>
      <Heading level="3">Contracts</Heading>
      {currentContract && (
        <Box direction="row" gap="small">
          <Text>Current Contract: {currentContract.id}</Text>
          <Button
            hoverIndicator
            icon={<Close />}
            onClick={() => handleResetContract()}
            plain
          />
        </Box>
      )}
      {contracts.map(contract => (
        <Box direction="row" key={contract.id} gap="small">
          <Text weight="bold">Item Type</Text>
          <Text>{contract.itemType}</Text>
          <Text weight="bold">Volume</Text>
          <Text>{contract.volume}</Text>
          <Text weight="bold">Value</Text>
          <Text>{contract.value}</Text>
          <Text weight="bold">Destination</Text>
          <Text>{contract.destination.name}</Text>
          <Button
            disabled={currentContract !== null}
            hoverIndicator
            icon={<Checkmark />}
            onClick={e =>
              handleSetContract(
                contracts.find(
                  contract => contract.id === e.target.parentElement.value
                )
              )
            }
            plain
            value={contract.id}
          />
        </Box>
      ))}
    </Box>
  )
}

ContractsDisplay.propTypes = {
  contracts: PropTypes.array.isRequired,
  currentContract: PropTypes.object,
  handleResetContract: PropTypes.func.isRequired,
  handleSetContract: PropTypes.func.isRequired
}

const mapStateToProps = ({ user, world }) => ({
  contracts: world.contracts,
  currentContract: user.contract
})

const mapDispatchToProps = dispatch => ({
  handleResetContract: () => {
    dispatch(resetContract())
  },
  handleSetContract: contract => {
    dispatch(setContract(contract))
  }
})

export default connect(
  mapStateToProps,
  mapDispatchToProps
)(ContractsDisplay)
