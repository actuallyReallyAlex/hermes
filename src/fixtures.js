export const defaultState = {
  ship: {
    cargo: {
      items: [],
      volumeRemaining: 5
    },
    location: {
      name: null,
      value: null
    }
  },
  ui: {
    view: 'Ship'
  },
  user: {
    cash: 100
  },
  world: {
    isTimerRunning: false,
    planets: [
      {
        id: "0",
        isHomePlanet: true,
        location: 0,
        items: [
          {
            name: 'Test Item',
            space: 1,
            value: 1,
            id: '0',
            destination: {
              name: 'Test Planet 2',
              value: 50
            }
          },
          {
            name: 'Test Item',
            space: 1,
            value: 1,
            id: '1',
            destination: {
              name: 'Test Planet 2',
              value: 50
            }
          },
          {
            name: 'Test Item',
            space: 1,
            value: 1,
            id: '2',
            destination: {
              name: 'Test Planet 2',
              value: 50
            }
          },
          {
            name: 'Test Item',
            space: 1,
            value: 1,
            id: '3',
            destination: {
              name: 'Test Planet 2',
              value: 50
            }
          },
          {
            name: 'Test Item',
            space: 1,
            value: 1,
            id: '4',
            destination: {
              name: 'Test Planet 2',
              value: 50
            }
          }
        ],
        name: 'Test Planet 1'
      },
      {
        id: "1",
        isHomePlanet: false,
        location: 50,
        items: [
          {
            name: 'Test Item',
            space: 1,
            value: 1,
            id: '5',
            destination: {
              name: 'Test Planet 1',
              value: 0
            }
          },
          {
            name: 'Test Item',
            space: 1,
            value: 1,
            id: '6',
            destination: {
              name: 'Test Planet 1',
              value: 0
            }
          },
          {
            name: 'Test Item',
            space: 1,
            value: 1,
            id: '7',
            destination: {
              name: 'Test Planet 1',
              value: 0
            }
          },
          {
            name: 'Test Item',
            space: 1,
            value: 1,
            id: '8',
            destination: {
              name: 'Test Planet 1',
              value: 0
            }
          },
          {
            name: 'Test Item',
            space: 1,
            value: 1,
            id: '9',
            destination: {
              name: 'Test Planet 1',
              value: 0
            }
          }
        ],
        name: 'Test Planet 2'
      },
      {
        id: "2",
        isHomePlanet: false,
        location: 100,
        items: [
          {
            name: 'Test Item',
            space: 1,
            value: 1,
            id: '10',
            destination: {
              name: 'Test Planet 2',
              value: 50
            }
          },
          {
            name: 'Test Item',
            space: 1,
            value: 1,
            id: '11',
            destination: {
              name: 'Test Planet 2',
              value: 50
            }
          },
          {
            name: 'Test Item',
            space: 1,
            value: 1,
            id: '12',
            destination: {
              name: 'Test Planet 2',
              value: 50
            }
          },
          {
            name: 'Test Item',
            space: 1,
            value: 1,
            id: '13',
            destination: {
              name: 'Test Planet 2',
              value: 50
            }
          },
          {
            name: 'Test Item',
            space: 1,
            value: 1,
            id: '14',
            destination: {
              name: 'Test Planet 2',
              value: 50
            }
          }
        ],
        name: 'Test Planet 3'
      }
    ]
  }
}
