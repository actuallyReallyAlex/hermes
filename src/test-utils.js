import React from 'react'
import configureStore from 'redux-mock-store'
import { Provider } from 'react-redux'
import { defaultState } from './fixtures'
import { render } from '@testing-library/react'

const mockThunks = ({ dispatch, getState }) => next => action => {
  const mockedAction = () => ({ type: 'MOCKED ACTION' })

  let returnValue

  if (typeof action === 'function' || action === undefined) {
    returnValue = mockedAction
  } else {
    returnValue = next(action)
  }

  return returnValue
}

const middlewares = [mockThunks]

export const mockStore = configureStore(middlewares)

export const customRender = options => {
  const { component, state, props } = options

  const Component = component

  return render(
    <Provider store={mockStore(state ? state : defaultState)}>
      <Component {...props} />
    </Provider>
  )
}
