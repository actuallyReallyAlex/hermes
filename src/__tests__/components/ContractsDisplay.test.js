import { customRender } from '../../test-utils'
import ContractsDisplay from '../../components/ContractsDisplay'
import { defaultState } from '../../fixtures'
import { fireEvent } from '@testing-library/dom';

const customState = {
  ...defaultState,
  user: {
    ...defaultState.user,
    contracts: [{ id: '0', itemType: 'Plasma', value: 2, volume: 1 }]
  }
}

describe('<ContractsDisplay />', () => {
  it('Should render the <ContractsDisplay /> component.', () => {
    const container = customRender({
      component: ContractsDisplay,
      state: customState
    })
    expect(container.asFragment()).toMatchSnapshot()
  })

  it('Should handle selecting a contract.', () => {
    const { debug, container } = customRender({
      component: ContractsDisplay,
      state: customState
    })

    const button = container.querySelector('button')

    fireEvent.click(button)
  })
})
