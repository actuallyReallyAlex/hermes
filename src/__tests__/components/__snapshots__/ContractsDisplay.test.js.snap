// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<ContractsDisplay /> Should render the <ContractsDisplay /> component. 1`] = `
<DocumentFragment>
  <div
    class="StyledBox-sc-13pk1d4-0 laNWCK"
  >
    <div
      class="StyledBox-sc-13pk1d4-0 IRSNj"
    >
      <h3
        class="StyledHeading-sc-1rdh4aw-0 ftOecI"
      >
        Contracts
      </h3>
      <div
        class="StyledBox-sc-13pk1d4-0 LPrfA"
      >
        <button
          class="StyledButton-sc-323bzc-0 lnKzGb"
          type="button"
        >
          Create
        </button>
        <button
          class="StyledButton-sc-323bzc-0 lnKzGb"
          type="button"
        >
          View
        </button>
      </div>
    </div>
  </div>
</DocumentFragment>
`;
